pulsar/consumer_impl.go

```diff
@@ -374,6 +374,7 @@ func (c *consumer) Close() {
 		}
 		wg.Wait()
 		close(c.closeCh)
+		close(c.messageCh)
```

Comment:
If the user passes in a channel and the channel is shared close could be called multiple times and there will be a panic.

Reply:
Yes. I already found this problem. It is hard to decide where to close the message channel.

Reply:
Maybe I should close message channel in multiTopicConsumer.Close() and  regexConsumer.Close(), and remove it in internal consumer.

Reply:
> Maybe I should close message channel in multiTopicConsumer.Close() and regexConsumer.Close(), and remove it in internal consumer.
 
Done. Commit again

Reply:
If a users passes a message channel in the client should not close that channel.

Reply:
I just came across this issue, and allowing the user to pass in a message channel seems like a mistake to me, because it inverts control. If the user wants to multiplex messages from different consumers onto a channel they control, it should be up to them to handle that with a select clause, [runtime.Select](https://pkg.go.dev/reflect#Select), or a combination of concurrency primitives in their code.

Reply:
> I just came across this issue, and allowing the user to pass in a message channel seems like a mistake to me, because it inverts control. If the user wants to multiplex messages from different consumers onto a channel they control, it should be up to them to handle that with a select clause, [runtime.Select](https://pkg.go.dev/reflect#Select), or a combination of concurrency primitives in their code.

Yes, I agree that this is a bad idea for users to pass their message channel. We should remove it in the Go client 1.0.0. This feature makes it difficult to develop and fix the consumer.